junebug/amqp.py

```diff
@@ -0,0 +1,153 @@
+        log.err("AmqpFactory connection failed (%s)" % (
+            reason.getErrorMessage(),))
+        super(AmqpFactory, self).clientConnectionFailed(connector, reason)
+
+    def clientConnectionLost(self, connector, reason):
+        log.err("AmqpFactory client connection lost (%s)" % (
+            reason.getErrorMessage(),))
+        self.disconnected_callback()
+        super(AmqpFactory, self).clientConnectionLost(connector, reason)
+
+
+class RoutingKeyError(Exception):
+    def __init__(self, value):
+        self.value = value
+
+    def __str__(self):
+        return repr(self.value)
+
+
+class JunebugAMQClient(AMQClient, object):
+    exchange_name = "vumi"
+    routing_key = "routing_key"
+    delivery_mode = 2  # save to disk
+
+    @inlineCallbacks
+    def connectionMade(self):
+        super(JunebugAMQClient, self).connectionMade()
+        yield self.authenticate(self.factory.amqp_config['username'],
+                                self.factory.amqp_config['password'])
+        # authentication was successful
+        log.msg("Got an authenticated AMQP connection")
+        self.factory.connected_callback(self)
+
+    @inlineCallbacks
+    def get_channel(self):
+        """If channel is None a new channel is created"""
+        if not hasattr(self, 'cached_channel'):
+            channel_id = self.get_new_channel_id()
+            channel = yield self.channel(channel_id)
+            yield channel.channel_open()
+            self.cached_channel = channel
+        else:
+            channel = self.cached_channel
+        returnValue(channel)
+
+    def get_new_channel_id(self):
+        """
+        AMQClient keeps track of channels in a dictionary. The
+        channel ids are the keys, get the highest number and up it
+        or just return zero for the first channel
+        """
+        return (max(self.channels) + 1) if self.channels else 0
+
+    def check_routing_key(self, routing_key):
+        if(routing_key != routing_key.lower()):
+            raise RoutingKeyError("The routing_key: %s is not all lower case!"
+                                  % (routing_key))
+
+    def publish_message(self, message, **kwargs):
+        d = self.publish_raw(message.to_json(), **kwargs)
+        d.addCallback(lambda r: message)
+        return d
+
+    def publish_raw(self, data, **kwargs):
+        amq_message = Content(data)
+        amq_message['delivery mode'] = kwargs.pop(
+            'delivery_mode', self.delivery_mode)
+        return self.publish(amq_message, **kwargs)
+
+    @inlineCallbacks
+    def publish(self, message, **kwargs):
+        exchange_name = kwargs.get('exchange_name') or self.exchange_name
+        routing_key = kwargs.get('routing_key') or self.routing_key
+        self.check_routing_key(routing_key)
+        channel = yield self.get_channel()
+        yield channel.basic_publish(
+            exchange=exchange_name, content=message, routing_key=routing_key)
```

Comment:
Would it maybe be worth testing that this class's methods are doing the right thing? I think since testing via real rabbit would be more work that it is worth, stubbing `basic_publish()` and `channel()` might be enough to test that these methods are doing the right thing. Sorry for commenting on this so late, it should've been done in the first review.